### PR TITLE
Forbid unsafe in most crates in the engine

### LINKED ISSUE #3824: Deny unsafe code in most of the code base
## What problem does this solve or what need does it fill?

While undeniably useful and often irreplaceable, unsafe code is inherently challenging to audit and maintain.

Uses of unsafe code outside of the core engine crate should not be necessary; safe APIs for missing functionality should be exposed instead.

## What solution would you like?

Add `#![deny(unsafe_code)]` annotations to most bevy crates at the top of lib.rs.

There are a few clear exceptions:
- `bevy_ecs`
  - some submodules, like events, should probably forbid unsafe code
- `bevy_crevice`
- `bevy_derive`
- `bevy_reflect`

There are some surprising usages of `unsafe` that should be handled in one form or another.
- `bevy_render`: used for working with raw window handles
- `bevy_dynamic_plugin` + `bevy_app`
  - this functionality is, as far as I'm aware, entirely unused
- `bevy_core`: used for fixed timestep run criteria...
- `bevy_transform`: used for working with entity hierarchies (#3742 please...)
- `bevy_winit` and `bevy_ui`: used to manually impl Send and Sync for external types
- `bevy_tasks`: used to [transmute an executor lifetime to 'static](https://github.com/bevyengine/bevy/blob/e30199f7a9cf6429aed6c0b6f66fe76b791e90aa/crates/bevy_tasks/src/task_pool.rs#L175)
- `bevy_log`: used to [log messages for Android](https://github.com/bevyengine/bevy/blob/5127274b3120152f18b0b1fb1cd31fb159dac126/crates/bevy_log/src/android_tracing.rs)

Each of these should be rewritten if feasible, or explicitly allowed if not.

##  Alternatives

We could use `#![forbid(unsafe_code)]` instead, but the ability to allow unsafe in carefully thought-out one-off instances is probably too valuable for many of these crates.

We could use #![warn(unsafe_code)] instead: keeping it at warning won't be a bother when you want to try things out locally, and Bevy CI will deny warnings.

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Resolves #3824.  `unsafe` code should be the exception, not the norm in Rust. It's obviously needed for various use cases as it's interfacing with platforms and essentially running the borrow checker at runtime in the ECS, but the touted benefits of Bevy is that we are able to heavily leverage Rust's safety, and we should be holding ourselves accountable to that by minimizing our unsafe footprint.

## Solution
Deny `unsafe_code` workspace wide. Add explicit exceptions for the following crates, and forbid it in almost all of the others.

 * bevy_ecs - Obvious given how much unsafe is needed to achieve performant results
 * bevy_ptr - Works with raw pointers, even more low level than bevy_ecs.
 * bevy_render - due to needing to integrate with wgpu
 * bevy_window - due to needing to integrate with raw_window_handle
 * bevy_utils - Several unsafe utilities used by bevy_ecs. Ideally moved into bevy_ecs instead of made publicly usable.
 * bevy_reflect - Required for the unsafe type casting it's doing.
 * bevy_transform - for the parallel transform propagation
 * bevy_gizmos  - For the SystemParam impls it has.
 * bevy_assets - To support reflection. Might not be required, not 100% sure yet.
 * bevy_mikktspace - due to being a conversion from a C library. Pending safe rewrite.
 * bevy_dynamic_plugin - Inherently unsafe due to the dynamic loading nature.

Several uses of unsafe were rewritten, as they did not need to be using them:

 * bevy_text - a case of `Option::unchecked` could be rewritten as a normal for loop and match instead of an iterator.
 * bevy_color - the Pod/Zeroable implementations were replaceable with bytemuck's derive macros.

### CODE REVIEW

**Review by bushrat011899 — APPROVED** [review]
Definitely a good change. Ideally, Bevy would contain zero `unsafe` code. While that may be impossible, it's still worth minimising the risk.

File: crates/bevy_dynamic_plugin/src/loader.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#![allow(unsafe_code)]
```
Comment (alice-i-cecile) [inline crates/bevy_dynamic_plugin/src/loader.rs:1]: 😂 No targeted allows here.

File: crates/bevy_text/src/pipeline.rs [hunk 2]
```diff
@@ -129,32 +129,25 @@ impl TextMeasureInfo {
...
-                return Err(TextError::NoSuchFont);
-            }
-        }
-        let (auto_fonts, sections) = sections
-            .iter()
-            .enumerate()
-            .map(|(i, section)| {
-                // SAFETY: we exited early earlier in this function if
-                // one of the fonts was missing.
-                let font = unsafe { fonts.get(&section.style.font).unwrap_unchecked() };
```
Comment (alice-i-cecile) [inline crates/bevy_text/src/pipeline.rs:143]: Good choice, this is the sort of needless unsafe I was hoping to clean up.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@james7132 @JMS55 @rodolphito Meshlets added new unsafe code that will need to be allowed to get this mergeable.

**Comment by james7132** [comment]
@JMS55 is there a reason why PersistentGpuBufferable is an unsafe trait? Both of it's functions are not used in an unsafe context without existing safety checks.

**Comment by JMS55** [comment]
Iirc it's because there's no checks for it, but the data has to be 4 byte aligned (or whatever wgpu wants), and it needs to accurately report the size in one function that then gets written in the second.

**Comment by james7132** [comment]
If that's the case, the APIs on wgpu's end need to be unsafe, which I don't think is the case. As far as CPU-side memory safety goes, this looks safe to me, even if the implementation of the trait is wrong.

**Review by BD103 — APPROVED** [review]
Looks good! What's your reason for `forbid`-ing many of the crates when they already inherit `deny` from the workspace?

**Comment by alice-i-cecile** [comment]
Merging is blocked on https://github.com/bevyengine/bevy/issues/12728

**Comment by james7132** [comment]
Allowing it for now and using #12743 to address the soundness issues.

**Pull request merged**